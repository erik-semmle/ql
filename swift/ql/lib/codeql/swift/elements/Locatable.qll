private import codeql.swift.generated.Locatable
private import codeql.swift.elements.File

class Locatable extends LocatableBase {
  pragma[nomagic]
  override Location getLocation() {
    result = LocatableBase.super.getLocation()
    or
    not exists(LocatableBase.super.getLocation()) and result instanceof UnknownLocation
  }

  /**
   * Gets the primary file where this element occurs.
   */
  File getFile() { result = getLocation().getFile() }
}
