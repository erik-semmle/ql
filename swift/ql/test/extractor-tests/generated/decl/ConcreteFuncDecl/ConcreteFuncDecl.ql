// generated by codegen/codegen.py
import codeql.swift.elements
import TestUtils

from
  ConcreteFuncDecl x, string hasSelfParam, int getNumberOfParams, string hasBody,
  int getNumberOfGenericTypeParams, ModuleDecl getModule, Type getInterfaceType, string getName
where
  toBeTested(x) and
  not x.isUnknown() and
  (if x.hasSelfParam() then hasSelfParam = "yes" else hasSelfParam = "no") and
  getNumberOfParams = x.getNumberOfParams() and
  (if x.hasBody() then hasBody = "yes" else hasBody = "no") and
  getNumberOfGenericTypeParams = x.getNumberOfGenericTypeParams() and
  getModule = x.getModule() and
  getInterfaceType = x.getInterfaceType() and
  getName = x.getName()
select x, "hasSelfParam:", hasSelfParam, "getNumberOfParams:", getNumberOfParams, "hasBody:",
  hasBody, "getNumberOfGenericTypeParams:", getNumberOfGenericTypeParams, "getModule:", getModule,
  "getInterfaceType:", getInterfaceType, "getName:", getName
