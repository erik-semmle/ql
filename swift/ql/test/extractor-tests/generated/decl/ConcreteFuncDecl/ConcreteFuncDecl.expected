| functions.swift:1:1:3:1 | foo() | hasSelfParam: | no | getNumberOfParams: | 0 | hasBody: | yes | getNumberOfGenericTypeParams: | 0 | getModule: | file://:0:0:0:0 | functions | getInterfaceType: | () -> Int | getName: | foo() |
| functions.swift:5:1:7:1 | bar(_:d:) | hasSelfParam: | no | getNumberOfParams: | 2 | hasBody: | yes | getNumberOfGenericTypeParams: | 0 | getModule: | file://:0:0:0:0 | functions | getInterfaceType: | (Int, Double) -> Int | getName: | bar(_:d:) |
| functions.swift:10:5:10:28 | noBody(x:) | hasSelfParam: | yes | getNumberOfParams: | 1 | hasBody: | no | getNumberOfGenericTypeParams: | 0 | getModule: | file://:0:0:0:0 | functions | getInterfaceType: | <Self where Self : Beep> (Self) -> (Int) -> Int | getName: | noBody(x:) |
| functions.swift:13:1:15:1 | variadic(_:) | hasSelfParam: | no | getNumberOfParams: | 1 | hasBody: | yes | getNumberOfGenericTypeParams: | 0 | getModule: | file://:0:0:0:0 | functions | getInterfaceType: | (Int...) -> () | getName: | variadic(_:) |
| functions.swift:17:1:19:1 | generic(x:y:) | hasSelfParam: | no | getNumberOfParams: | 2 | hasBody: | yes | getNumberOfGenericTypeParams: | 2 | getModule: | file://:0:0:0:0 | functions | getInterfaceType: | <X, Y> (x: X, y: Y) -> () | getName: | generic(x:y:) |
